packages/react/src/components/PaginationNav/PaginationNav.js

```diff
@@ -0,0 +1,363 @@
+/**
+ * Copyright IBM Corp. 2020
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import classnames from 'classnames';
+import {
+  CaretRight16,
+  CaretLeft16,
+  OverflowMenuHorizontal16,
+} from '@carbon/icons-react';
+import { settings } from 'carbon-components';
+import Button from '../Button';
+
+const { prefix } = settings;
+
+const getCuts = (totalItems, itemsThatFit, page) => {
+  if (itemsThatFit >= totalItems) {
+    return {
+      front: 0,
+      back: 0,
+    };
+  }
+
+  let reservedSpots = 4; // first + overflow + overflow + last
+
+  // remove reserveration for first item when only 4 items can be displayed
+  if (itemsThatFit < 5) reservedSpots -= 1;
+  // remove one overflow reservation when current page would hide it
+  if (page < 2 || page > totalItems - 3) reservedSpots -= 1;
+
+  const shownItems = itemsThatFit - reservedSpots;
+  const hiddenItems = totalItems - shownItems - (itemsThatFit < 5 ? 1 : 2);
+
+  let frontHidden = Math.floor(hiddenItems / 2);
+  let backHidden = Math.ceil(hiddenItems / 2);
+
+  if (itemsThatFit > 4) {
+    if (page <= frontHidden) {
+      // if current page would be in the front overflow
+      let newFrontHidden = frontHidden - (frontHidden - page) - 1;
+      newFrontHidden = newFrontHidden >= 0 ? newFrontHidden : 0;
+      const delta = frontHidden - newFrontHidden;
+
+      frontHidden -= delta;
+      backHidden += delta;
+    } else if (page >= totalItems - backHidden - 1) {
+      // if the current page would be in the back overflow
+      let newBackHidden = totalItems - 2 - page;
+      newBackHidden = newBackHidden >= 0 ? newBackHidden : 0;
+      const delta = backHidden - newBackHidden;
+
+      frontHidden += delta;
+      backHidden -= delta;
+    }
+  } else {
+    frontHidden = page > 1 ? page : 0;
+    frontHidden = frontHidden <= totalItems - 3 ? frontHidden : totalItems - 3;
+
+    backHidden = page < totalItems - 2 ? totalItems - page - 2 : 0;
+    backHidden = backHidden <= totalItems - 3 ? backHidden : totalItems - 3;
+  }
+
+  return {
+    front: frontHidden,
+    back: backHidden,
+  };
+};
+
+export default class PaginationNav extends Component {
```

Comment:
Quick style-nit, for net-new code we try and use function components with hooks 👀